plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/PipelineExecutionService.java

```diff
@@ -40,22 +42,46 @@ public PipelineExecutionService(PipelineStore store, ThreadPool threadPool) {
         this.threadPool = threadPool;
     }
 
-    public void execute(IngestDocument ingestDocument, String pipelineId, Listener listener) {
+    public void execute(IndexRequest indexRequest, String pipelineId, Listener listener) {
         Pipeline pipeline = store.get(pipelineId);
         if (pipeline == null) {
             listener.failed(new IllegalArgumentException("pipeline with id [" + pipelineId + "] does not exist"));
             return;
         }
 
-        threadPool.executor(THREAD_POOL_NAME).execute(new Runnable() {
-            @Override
-            public void run() {
-                try {
-                    pipeline.execute(ingestDocument);
-                    listener.executed(ingestDocument);
-                } catch (Throwable e) {
-                    listener.failed(e);
+        threadPool.executor(THREAD_POOL_NAME).execute(() -> {
+            String index = indexRequest.index();
+            String type = indexRequest.type();
+            String id = indexRequest.id();
+            String routing = indexRequest.routing();
+            String parent = indexRequest.parent();
+            String timestamp = indexRequest.timestamp();
+            String ttl = null;
+            if (indexRequest.ttl() != -1) {
+                ttl = String.valueOf(indexRequest.ttl());
+            }
+            Map<String, Object> sourceAsMap = indexRequest.sourceAsMap();
+            IngestDocument ingestDocument = new IngestDocument(index, type, id, routing, parent, timestamp, ttl, sourceAsMap);
+            try {
+                pipeline.execute(ingestDocument);
+                if (ingestDocument.isSourceModified()) {
+                    indexRequest.source(ingestDocument.getSource());
+                }
+                if (ingestDocument.isMetaDataModified()) {
+                    indexRequest.index(ingestDocument.getMetadata(IngestDocument.MetaData.INDEX));
+                    indexRequest.type(ingestDocument.getMetadata(IngestDocument.MetaData.TYPE));
+                    indexRequest.id(ingestDocument.getMetadata(IngestDocument.MetaData.ID));
+                    indexRequest.routing(ingestDocument.getMetadata(IngestDocument.MetaData.ROUTING));
+                    indexRequest.parent(ingestDocument.getMetadata(IngestDocument.MetaData.PARENT));
+                    indexRequest.timestamp(ingestDocument.getMetadata(IngestDocument.MetaData.TIMESTAMP));
+                    String ttlStr = ingestDocument.getMetadata(IngestDocument.MetaData.TTL);
+                    if (ttlStr != null) {
+                        indexRequest.ttl(Long.parseLong(ttlStr));
```

Comment:
should we validate that the ttl is a valid long as part of the metadata processor? We could add a small validate method to the enum, which does nothing by default but that the ttl one implements.